### PR TITLE
Implementing Push and Request Logic

### PR DESCRIPTION
**Pull request by chadlagore**
**Status** _open for visibility._

# What Changed

+ Added `Worker`, `Work`, and a dispatcher.
+ Currently writing tests.

## How to Test Drive

+ Currently `go test`, more to come...

### CODE REVIEW

File: app/app.go [hunk 1]
```diff
@@ -36,9 +44,12 @@ func Run(cfg conf.Config) {
 	log.Infof("Starting listener on %s:%d", cfg.Interface, cfg.Port)
 	peer.LocalAddr = fmt.Sprintf("%s:%d", cfg.Interface, cfg.Port)
 	go func() {
-		err := conn.Listen(fmt.Sprintf("%s:%d", cfg.Interface, cfg.Port), peer.ConnectionHandler)
+		address := fmt.Sprintf("%s:%d", cfg.Interface, cfg.Port)
```
Comment (bfbachmann) [inline app/app.go:47]: Might as well move this line up above this goroutine and use `address` in the calls to `log.Infof` on line 44 and `Fatalf` on like 52.

File: app/app.go [hunk 2]
```diff
@@ -63,20 +74,34 @@ func Run(cfg conf.Config) {
...
+	// Request the blockchain.
+	log.Info("Requesting blockchain from peers... ")
+	RequestBlockChain()
+
+	// Return to command line.
 	select {}
 }
 
-// RequestHandler is called every time a peer sends us a request message expect
+// RequestHandler is called every time a peer sends us a request message except
```
Comment (bfbachmann) [inline app/app.go]: sad

File: app/work.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
+package app
+
+import "github.com/ubclaunchpad/cumulus/blockchain"
+
+const (
+	// BlockQueueBuffer is the size of the BlockQueue channel.
+	BlockQueueBuffer = 100
```
Comment (bfbachmann) [inline app/work.go]: I would call this BlockQueueSize. Same for TransactionQueueBuffer.

**Review by bfbachmann — COMMENTED** [review]
Looks good, I just don't get why we need to have distributed workers and a job dispatcher. Won't these jobs just take a few seconds at most? I figured we could implement the same functionality with just a simple PushHandler and RequestHandler that do what HandleTransaction and HandleBock do without needing to worry about queueing jobs and creating workers etc.

**Comment by chadlagore** [comment]
@bfbachmann We should benchmark it, but `ValidTransaction` and `ValidBlock` are actually quite expensive, especially if we choose to not cap the block size. I see your point, but I'm not convinced that this is premature optimization yet.

File: app/app.go [hunk 4]
```diff
@@ -85,17 +111,56 @@ func RequestHandler(req *msg.Request) msg.Response {
...
 	case msg.ResourceTransaction:
+		work := TransactionWork{}
+		TransactionWorkQueue <- work
 	default:
 		// Invalid resource type. Ignore
 	}
+}
+
+// Shutdown shuts down the node (save the pool and the chain to disk).
+func Shutdown() {}
```
Comment (jordanschalm) [inline app/app.go]: Saving to disk might not be a prototype feature

File: app/app.go [hunk 5]
```diff
@@ -85,17 +111,56 @@ func RequestHandler(req *msg.Request) msg.Response {
...
+// incoming messages.
+func initializeNode() {
+	tpool = pool.New()
+	intializeQueues()
+	initializeWorkers()
+}
+
+// intializeQueues makes all necessary queues.
+func intializeQueues() {
+	BlockWorkQueue = make(chan BlockWork, BlockQueueBuffer)
```
Comment (jordanschalm) [inline app/app.go]: Maybe `BlockBufferSize` or something to make it more clear that it's a size

File: app/work.go [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+	// TransactionQueueBuffer is the size of the BlockQueue channel.
+	TransactionQueueBuffer = 100
+)
+
+// Responder is used to handle requests who require a response.
+type Responder interface {
+	Send(ok bool)
+	Lock()
```
Comment (jordanschalm) [inline app/work.go:15]: Why lock/unlock? Why does send only take a bool?

File: app/worker.go [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+func (w GenericWorker) Start() {
+	go func() {
+		for {
+			// Wait for work.
+			log.WithFields(log.Fields{
+				"id": w.ID,
+			}).Debug("Worker waiting for work.")
+			select {
+			case work := <-TransactionWorkQueue:
+				log.WithFields(log.Fields{
```
Comment (jordanschalm) [inline app/worker.go]: You can avoid this by adding a `log` field to the worker and using 
```
worker := Worker{
  log: log.WithField("id", id),
  ...
}
```

Comment (jordanschalm) [reply app/worker.go]: The field will persist on every subsequent `log.Whatever` call

File: app/worker.go [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+				return
+			}
+		}
+	}()
+}
+
+// NB: We're currently imposing a validation layer at the app level
+// using methods like SetUnsafe to get the transaction into the
+// pool, etc. Figure out a nicer way to do this, ie: where *should*
+// we validate?
```
Comment (jordanschalm) [inline app/worker.go]: I thought `pool` would have a channel of incoming, unvalidated transactions that `app` could push onto. `pool` should do all *transaction* validation

File: app/worker.go [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+// HandleTransaction handles new instance of TransactionWork.
+func (w *GenericWorker) HandleTransaction(work TransactionWork) {
+	ok, _ := chain.ValidTransaction(work.Transaction)
+	if ok {
+		tpool.SetUnsafe(work.Transaction)
+	}
+
+	// Respond to the request if a response method was provided.
+	if work.Responder != nil {
+		work.Responder.Lock()
```
Comment (jordanschalm) [inline app/worker.go:66]: Minor thing but idiom is:
```
x.Lock()
defer x.Unlock()
...
```

Rationale is if the body ever becomes more complex your `defer` will handle it without needing to think about it

**Review by jordanschalm — APPROVED** [review]
Looks good. Couple of minor comments. It does seem quite complex, but maybe that will be necessary.

File: app/work.go (hunk 6 above)
Comment (chadlagore) [reply app/work.go:15]: `Send` is sort of prototype-y. Not sure what message may need to go back to the app.

Comment (chadlagore) [reply app/work.go:15]: Threads may share a `Responder`.

File: app/worker.go (hunk 7 above)
Comment (chadlagore) [reply app/worker.go]: WHAT

File: app/worker.go (hunk 8 above)
Comment (chadlagore) [reply app/worker.go]: I think so too. I'm would be doing transaction validation at the `pool` level and block validation at the `app` level, it ust felt weird. I'll change it back.

File: app/worker.go (hunk 9 above)
Comment (chadlagore) [reply app/worker.go:66]: Yeah, I'm game thx.

File: app/app.go (hunk 5 above)
Comment (chadlagore) [reply app/app.go]: Totally 🤦‍♂️

File: app/work.go (hunk 3 above)
Comment (chadlagore) [reply app/work.go]: Totally 🤦‍♀️

File: app/app.go (hunk 2 above)
Comment (chadlagore) [reply app/app.go]: ✅

File: app/app.go (hunk 1 above)
Comment (chadlagore) [reply app/app.go:47]: I think `Run` will get a big overhaul yet.

File: app/worker.go (hunk 7 above)
Comment (jordanschalm) [reply app/worker.go]: Just need to do `worker.log.Whatever` rather than `log.Whatever`

File: app/worker.go (hunk 8 above)
Comment (chadlagore) [reply app/worker.go]: I wonder if it makes sense to have the blockchain make those choices too. ie `AppendBlock` simply fails if the block is invalid.

Comment (chadlagore) [reply app/worker.go]: It makes no sense to append a non-legit block.

**Pull request merged**